File: cpp/DataStructures/Queue_test.cpp
#include <iostream>
#include "Queue.h" // Assuming the queue implementation is in queue.h
using namespace std;

void testQueue() {
    Queue q;

    // Test: Enqueue elements
    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);
    cout << "Peek after enqueues: " << q.peek() << " (Expected: 10)" << endl;

    // Test: Dequeue elements
    q.dequeue();
    cout << "Peek after first dequeue: " << q.peek() << " (Expected: 20)" << endl;

    q.dequeue();
    cout << "Peek after second dequeue: " << q.peek() << " (Expected: 30)" << endl;

    q.dequeue();
    cout << "Is queue empty after all dequeues? " << (q.isEmpty() ? "Yes" : "No") << " (Expected: Yes)" << endl;

    // Test: Dequeue from empty queue
    q.dequeue(); // Should print error message

    // Test: Peek from empty queue
    cout << "Peek from empty queue: " << q.peek() << " (Expected: -1)" << endl;
}

int main() {
    testQueue();
    return 0;
}

File: go/DataStructures/Queue/Queue_test.go
package Queue

import (
	"testing"
	"github.com/stretchr/testify/assert"
)

func TestEnqueue(t *testing.T) {
	queue := NewQueue()

	queue.Enqueue(1)
	assert.Equal(t, 1, queue.Peek(), "Front of queue should be 1")

	queue.Enqueue(2)
	assert.Equal(t, 1, queue.Peek(), "Front of queue should remain 1 after enqueuing 2")
}

func TestDequeue(t *testing.T) {
	queue := NewQueue()
	queue.Enqueue(1)
	queue.Enqueue(2)

	err := queue.Dequeue()
	assert.Nil(t, err, "Dequeue should not return an error")
	assert.Equal(t, 2, queue.Peek(), "Front of queue should be 2 after dequeuing 1")

	err = queue.Dequeue()
	assert.Nil(t, err, "Dequeue should not return an error")
	assert.True(t, queue.IsEmpty(), "Queue should be empty after dequeuing all elements")

	err = queue.Dequeue()
	assert.NotNil(t, err, "Dequeue should return an error when the queue is empty")
}

func TestPeek(t *testing.T) {
	queue := NewQueue()
	queue.Enqueue(5)
	assert.Equal(t, 5, queue.Peek(), "Peek should return the front element")
}

func TestIsEmpty(t *testing.T) {
	queue := NewQueue()
	assert.True(t, queue.IsEmpty(), "New queue should be empty")

	queue.Enqueue(1)
	assert.False(t, queue.IsEmpty(), "Queue should not be empty after enqueuing an element")

	queue.Dequeue()
	assert.True(t, queue.IsEmpty(), "Queue should be empty after dequeuing all elements")
}

